solidity/contracts/vault/TBTCOptimisticMinting.sol

```diff
@@ -0,0 +1,112 @@
+// SPDX-License-Identifier: GPL-3.0-only
+
+// ██████████████     ▐████▌     ██████████████
+// ██████████████     ▐████▌     ██████████████
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+// ██████████████     ▐████▌     ██████████████
+// ██████████████     ▐████▌     ██████████████
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+
+pragma solidity 0.8.17;
+
+import "../bridge/Bridge.sol";
+import "../bridge/Deposit.sol";
+
+abstract contract TBTCOptimisticMinting {
+
+    // TODO: make it governable?
+    uint256 public constant optimisticMintingDelay = 3 hours;
+
+    Bridge public bridge;
+
+    mapping(address => bool) public isMinter;
+    mapping(address => bool) public isWatchman;
+
+    mapping(uint256 => uint256) public pendingOptimisticMints;
+
+    mapping(address => uint256) public optimisticMintingDebt;
+
+    modifier onlyMinter() {
+        require(isMinter[msg.sender], "Caller is not a minter");
+        _;
+    }
+
+    modifier onlyWatchman() {
+        require(isWatchman[msg.sender], "Caller is not a watchman");
+        _;
+    }
+
+    constructor(Bridge _bridge) {
+        require(
+            address(_bridge) != address(0),
+            "Bridge can not be the zero address"
+        );
+
+        bridge = _bridge;
+    }
+
+    function _mint(address minter, uint256 amount) virtual internal;
+
+    function optimisticMint(uint256 depositKey) external onlyMinter {
+        Deposit.DepositRequest memory deposit = bridge.deposits(depositKey);
+
+        require(deposit.revealedAt != 0, "The deposit has not been revealed");
+        require(deposit.sweptAt == 0, "The deposit is already swept");
+        require(deposit.vault == address(this), "Unexpected vault address");
+
+        pendingOptimisticMints[depositKey] = block.timestamp;
+
+        // TODO: emit an event
+    }
+
+    function finalizeOptimisticMint(uint256 depositKey) external onlyMinter {
+        uint256 requestedAt = pendingOptimisticMints[depositKey];
+        require(requestedAt != 0, "Optimistic minting not requested");
+        require(
+            block.timestamp - requestedAt > optimisticMintingDelay,
+            "Optimistic minting delay has not passed yet"
+        );
+
+        Deposit.DepositRequest memory deposit = bridge.deposits(depositKey);
+        require(deposit.sweptAt == 0, "The deposit is already swept");
+
+        // TODO: deal with the minting fee
+        _mint(deposit.depositor, deposit.amount);
+        optimisticMintingDebt[deposit.depositor] += deposit.amount;
+
+        delete pendingOptimisticMints[depositKey];
+
+        // TODO: emit an event
+    }
+
+    // TODO: Is this function convenient enough to block minting at 3AM ?
+    //       Do we want to give watchment a chance to temporarily disable
+    //       finalizeOptimisticMint ?
```

Comment:
Thinking of a 3 AM situation for some watchman to block minting. We could have a mechanism allowing to pause optimistic minting for 24h by any watchmen. This pause could be disabled by the owner of this contract. 

If the given watchman proves to be DoSing this way, the owner can remove the watchman from the permissioned set. 

What do you think?

Reply:
Let's say that one of the minters gets compromised and they start requesting mints in a loop. With just `cancelOptimisticMint`, watchmen would have to run the same loop and this is becoming a game of speed and who has more ether. Options are, for example:
- `pauseOptimisticMints` executed by one watchman, pausing all mints for 24h. The given watchman can pause mints only one time. The owner can `resumeOptimisticMints` at any time.
- `pauseOptimisticMints` executed by at least >= 2 watchman together, pausing all mints indefinitely, until the owner `resumeOptimisticMints`.

Reply:
`TBTCVault`'s owner can:
- [`initiateUpgrade(address _newVault) external onlyOwner`](https://github.com/keep-network/tbtc-v2/blob/8b9d2629bf4333e650a54f32a4da7cf86bf6785e/solidity/contracts/vault/TBTCVault.sol#L197)
- [`finalizeUpgrade() external onlyOwner onlyAfterUpgradeGovernanceDelay`](https://github.com/keep-network/tbtc-v2/blob/8b9d2629bf4333e650a54f32a4da7cf86bf6785e/solidity/contracts/vault/TBTCVault.sol#L212)
- [`recoverERC20FromToken(IERC20 token, address recipient, uint256 amount) external onlyOwner`](https://github.com/keep-network/tbtc-v2/blob/8b9d2629bf4333e650a54f32a4da7cf86bf6785e/solidity/contracts/vault/TBTCVault.sol#L230)
- [`recoverERC721FromToken(IERC721 token, address recipient, uint256 tokenId, bytes calldata data) external onlyOwner`](https://github.com/keep-network/tbtc-v2/blob/8b9d2629bf4333e650a54f32a4da7cf86bf6785e/solidity/contracts/vault/TBTCVault.sol#L244)
- [`recoverERC20(IERC20 token, address recipient, uint256 amount) external onlyOwner`](https://github.com/keep-network/tbtc-v2/blob/8b9d2629bf4333e650a54f32a4da7cf86bf6785e/solidity/contracts/vault/TBTCVault.sol#L260)
- [`recoverERC721(IERC721 token, address recipient, uint256 tokenId, bytes calldata data) external onlyOwner`](https://github.com/keep-network/tbtc-v2/blob/8b9d2629bf4333e650a54f32a4da7cf86bf6785e/solidity/contracts/vault/TBTCVault.sol#L274)

In other words, it has upgrade capabilities and emergency token recovery functionalities with a caveat that `recoverERC20` may also be used for withdrawing tBTC v1 if it's not withdrawn from v1<>v2 vending machine [before the upgrade](https://github.com/keep-network/tbtc-v2/blob/8b9d2629bf4333e650a54f32a4da7cf86bf6785e/solidity/contracts/bridge/VendingMachine.sol#L242).

This feels like a set of functionalities that should be assigned to the DAO/large multisig.

On the other hand, functions to remove an errant minter or unpause optimistic minting are not necessarily something we want to delegate to the DAO or large multisigs given the voting time.

We could introduce a `leadGuard` (?) role, assign it to a developer multisig, and organize ACLs in the following way:
- `addMinter onlyOwner` (DAO/council multisig)
- `removeMinter onlyLeadGuardOrOwner` (DAO/council multisig of developer multisig)
- `addGuard onlyLeadGuardOrOwner`
- `removeGuard onlyLeadGuardOrOwner`
- `unpauseMinting onlyLeadGuardOrOwner`
- `pauseMinting onlyGuard` - pauses until `unpauseMinting` is called

In the scenario when a minter gets compromised, any single guard can `pauseMinting`. Only the lead guard or owner (DAO) can unpause, previously removing the minter or guard.

Alternatively, we could not implement `pauseMinting` and just give any guard an option to `removeMinter`. The risk of DoS is the same as with the pause.

What do you think?

Reply:
If we were to introduce a `LeadGuard`, then why don't we make `pauseMinting` also callable by a `LeadGuard`, ie. `pauseMinting onlyLeadGuardOrGuard`?

Reply:
I approached the implementation and the idea with the lead guard is too complicated.

That's what I would do:
```
addMinter onlyOwner
removeMinter onlyOwnerOrGuardian
addGuardian onlyOwner
removeGuardian onlyOwner

pauseOptimisticMinting onlyOwner
unpauseOptimisticMinting onlyOwner
```

In case a Guardian gets compromised, it can remove all Minters and DoS the contract until the owner removes the compromised Guardian and re-adds Minters.

If Minter gets compromised, it can keep requesting minting for non-existing deposits, Guardians will have to cancel them and remove the errant Minter.

Pause functionality should be used for long-term pauses, not to deal with errant Minters.

Reply:
See 78d580c53b1057c9c07365e21f66fd3c7118738e and 9745ab844a70fc6a3b07b9500b6ac89141f98d42, please.